### PR TITLE
Update how towns/dungeons/gyms/routes requirements are handled

### PR DESCRIPTION
**Pull request by RedSparr0w**
All requirements should be updated,

Note:
Some will need more requirements added, as they weren't available before,
not sure which though, so will need to be done in a separate PR.

Edit:
Kanto requirements are all updated, some routes/dungeons/gyms are optional until specific points in the game.
Also had to update the Tutorial Quests as Viridian Forest is required before you reach Brock

### CODE REVIEW

File: src/scripts/worldmap/Routes.ts [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+    }
+
+    public static getRoutesByRegion(region: GameConstants.Region) {
+        return this.regionRoutes.filter(routeData => routeData.region == region);
+    }
+
+    public static normalizedNumber(region: GameConstants.Region, route: number): number {
+        return this.regionRoutes
+            .sort((routeA, routeB) => routeA.orderNumber - routeB.orderNumber)
+            .sort((routeA, routeB) => routeA.region - routeB.region)
```
Comment (Aegyo) [inline src/scripts/worldmap/Routes.ts]: Seems like something that will be running a lot, and return the same thing every time. Why don't we just do the sorting once after all routes have been added?

File: src/scripts/worldmap/MapHelper.ts [hunk 2]
```diff
@@ -205,40 +156,21 @@ class MapHelper {
...
-                    const routesList = routesNotCompleted.join(', ');
-                    if (routesNotCompleted.length > 1) {
-                        reqsList += `<br/>Routes ${routesList} still need to be completed.`;
-                    } else {
-                        reqsList += `<br/>Route ${routesList} still needs to be completed.`;
-                    }
+            town.requirements.forEach(requirement => {
+                if (requirement.isCompleted()) {
+                    return;
+                } else if (requirement instanceof GymBadgeRequirement) {
```
Comment (Aegyo) [inline src/scripts/worldmap/MapHelper.ts]: Not liking these `instanceof` checks which will need adding to if/when we add new Requirement types. Maybe Requirement should have a required `hint(): string` method we can call, and new requirements need only implement it?

**Review by Aegyo — COMMENTED** [review]
Haven't had a chance to play through the game on this branch yet, but here are the only issues I have after browsing all of the changes.

There are few requirements for things which seem redundant (eg, route 4 cleared and champion defeated for cerulean cave - route 4 cleared seems necessary to reach the champion), but it doesn't particularly bother me.

Also `EitherRequirement` suggests only two, but it can be many. Not sure if you want to bother renaming it to match `Array.some` or `Promise.any` (and a possible future `every`/`all`)

File: src/scripts/worldmap/MapHelper.ts (hunk 2 above)
Comment (RedSparr0w) [reply src/scripts/worldmap/MapHelper.ts]: Done!
Not sure if this is correct though, it doesn't seem to be required as an abstract.
Is there a better way?

File: src/scripts/worldmap/Routes.ts (hunk 1 above)
Comment (RedSparr0w) [reply src/scripts/worldmap/Routes.ts]: Good idea,
I've updated it to sort when a new route is added.

**Comment by RedSparr0w** [comment]
Changed from `EitherRequirement` -> `OneFromManyRequirement` not sure if that makes more sense, not sure what else to call it though.

Also removed some of the redundant requirements.

**Review by Aegyo — COMMENTED** [review]
After a quick play through, had a few thoughts:

When trying to access Cinnabar gym it tells me I am not worthy and should try completing more badges first, but the real reason I can't access is because I haven't done Pokemon Mansion. Can we update the gym access hint to use the Requirements for that gym? I could just do all of the orange things on the map to unlock it, but it would be nice to be told the exact reason I can't enter the gym.

Routes 47/48 can be accessed after Cianwood, but don't have an orderNumber to reflect this - is that on purpose?

Kanto had surf areas unlocked from both ends when the badge was unlocked, but Hoenn doesn't seem to do this. I would add route 109, 134, and New Mauville to the things unlocked by Petalburg gym, and also have 108 etc be opened from either side. 134 should not give access to 133 though, there are strong currents preventing that in the actual games.

Route 116 could do with an orderNumber, and possibly the water routes 105-109

**Comment by RedSparr0w** [comment]
I haven't done any of the requirement changes for Johto and Hoenn yet, only refactored them for the new format using the previous data.

> Can we update the gym access hint to use the Requirements for that gym? I could just do all of the orange things on the map to unlock it, but it would be nice to be told the exact reason I can't enter the gym.

Good idea, I completely forgot about those!

**Comment by Aegyo** [comment]
Ah okay, are the other region requirements going to be left for another PR, so that this can be merged?

**Comment by RedSparr0w** [comment]
Yeah might just update the other regions in the Sinnoh branch when I adjust that to fit this PR too.

**Comment by RedSparr0w** [comment]
Sweet, this should be ready for another review.

I've updated the Gyms to give the reason they cannot be accessed yet.

Also made it so the town won't show as orange if the Gym cannot be accessed yet.

Attached save pre Cinnabar Pokemon Mansion for testing:

[Pre Cinnabar Mansion.txt](https://github.com/pokeclicker-dev/pokeclicker/files/5074480/Pre.Cinnabar.Mansion.txt)

**Review by Aegyo — APPROVED** [review]
Looks good :+1:

**Pull request merged**